File: library/src/main/java/com/snail/commons/entity/CheckableParcelable.java
package com.snail.commons.entity;

import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;

/**
 * date: 2019/8/6 12:38
 * author: zengfansheng
 */
public class CheckableParcelable<T extends Parcelable> extends CheckableItem<T> implements Parcelable {
    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        Bundle bundle = new Bundle();
        bundle.putParcelable("items", getData());
        dest.writeBundle(bundle);
        dest.writeByte(isChecked() ? (byte) 1 : (byte) 0);
    }

    public CheckableParcelable() {
    }

    @SuppressWarnings("unchecked")
    protected CheckableParcelable(Parcel in) {
        Bundle bundle = in.readBundle(getClass().getClassLoader());
        if (bundle != null) {
            setData((T) bundle.getParcelable("items"));
        }
        setChecked(in.readByte() != 0);
    }

    public static final Parcelable.Creator<CheckableParcelable> CREATOR = new Parcelable.Creator<CheckableParcelable>() {
        @Override
        public CheckableParcelable createFromParcel(Parcel source) {
            return new CheckableParcelable(source);
        }

        @Override
        public CheckableParcelable[] newArray(int size) {
            return new CheckableParcelable[size];
        }
    };
}

File: library/src/main/java/com/snail/commons/entity/CheckableParcelable.kt
package com.snail.commons.entity

import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable

/**
 * 描述: 可选中并且实现Parcelable
 * 时间: 2018/9/7 09:58
 * 作者: zengfansheng
 */
class CheckableParcelable<T : Parcelable> : CheckableItem<T>, Parcelable {

    constructor()

    constructor(data: T?) : super(data)

    constructor(data: T?, isChecked: Boolean) : super(data, isChecked)

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        val bundle = Bundle()
        bundle.putParcelable(KEY_DATA, this.data)
        dest.writeBundle(bundle)
        dest.writeByte(if (isChecked) 1.toByte() else 0.toByte())
    }

    @Suppress("unchecked_cast")
    protected constructor(source: Parcel) {
        this.data = source.readBundle(javaClass.classLoader)?.getParcelable<Parcelable>(KEY_DATA) as? T
        this.isChecked = source.readByte().toInt() != 0
    }

    companion object {
        private const val KEY_DATA = "items"

        @JvmField
        val CREATOR: Parcelable.Creator<CheckableParcelable<*>> = object : Parcelable.Creator<CheckableParcelable<*>> {
            override fun createFromParcel(source: Parcel): CheckableParcelable<*> {
                return CheckableParcelable<Parcelable>(source)
            }

            override fun newArray(size: Int): Array<CheckableParcelable<*>?> {
                return arrayOfNulls(size)
            }
        }
    }
}
